Default protect from forgery

lugray opened the pull request:
### Summary

Fixes #29193.

Protect from forgery by default: Rather than protecting from forgery in the generated `ApplicationController`, add it to `ActionController::Base` by config. This configuration defaults to false to support older versions which have removed it from their `ApplicationController`, but is set to true for Rails 5.2.

Also adds ActionController::Base.skip_forgery_protection as a wrapper to `skip_before_action :verify_authenticity_token` for disabling forgery protection.

Commit:
Protect from forgery by default

Rather than protecting from forgery in the generated
ApplicationController, add it to ActionController::Base by config. This
configuration defaults to false to support older versions which have
removed it from their ApplicationController, but is set to true for
Rails 5.2.

lugray commented:
r? @rafaelfranca

rafaelfranca commented on actionpack/lib/action_controller/railtie.rb:
```diff
@@ -69,5 +69,13 @@ class Railtie < Rails::Railtie #:nodoc:
     end
+
+    initializer "action_controller.request_forgery_protection" do |app|
+      ActiveSupport.on_load(:action_controller) do
```
This is also going to run for `ActionController::Api`, so it is better to use `on_load(:action_controller_base` here. That way we can do:

```ruby
ActiveSupport.on_load(:action_controller_base) do
  if app.config.action_controller.default_protect_from_forgery
    protect_from_forgery with: :exception
  end
end
```

rafaelfranca commented on railties/lib/rails/application/configuration.rb:
```diff
@@ -96,6 +96,10 @@ def load_defaults(target_version)
           end
 
+          if respond_to?(:action_controller)
+            action_controller.default_protect_from_forgery = true
```
We need to add this config to the configurations.md guide.

lugray replied:
Done.

Edouard-chin commented on actionpack/lib/action_controller/metal/request_forgery_protection.rb:
```diff
@@ -128,6 +132,15 @@ def protect_from_forgery(options = {})
+      #
+      #   skip_before_action :verify_authenticity_token
+      #
+      # See skip_before_action for allowed options.
```
Small nitpicks, might use `+skip_before_action+` for [rdoc](https://docs.ruby-lang.org/en/2.1.0/RDoc/Markup.html#class-RDoc::Markup-label-Text+Markup)

lugray replied:
Thanks!

rafaelfranca reviewed (changes requested):
Last thing. Could you add a CHANGELOG entry?

eugeneius commented on railties/test/application/configuration_test.rb:
```diff
@@ -1209,6 +1209,15 @@ def create
+    test "config.action_controller.default_protect_from_forgery is true by default on production" do
+      app "production"
+
+      ActionController::Base.object_id # force lazy load hooks to run
```
This isn't necessary, since referencing `ActionController::Base ` on the next line will run the lazy load hooks. It's only needed when testing `ActionController::Parameters` options, as loading that class won't trigger the hooks.

eugeneius commented on railties/test/application/configuration_test.rb:
```diff
@@ -1209,6 +1209,15 @@ def create
       assert_equal false, ActionController::Parameters.action_on_unpermitted_parameters
     end
 
+    test "config.action_controller.default_protect_from_forgery is true by default on production" do
```
The behaviour here doesn't depend on the environment, so I'd ✂️ "on production" from the test name.

eugeneius commented on railties/test/application/configuration_test.rb:
```diff
@@ -1209,6 +1209,15 @@ def create
     end
 
+    test "config.action_controller.default_protect_from_forgery is true by default on production" do
+      app "production"
```
The rest of the tests in this file set this as "development" when the behaviour doesn't depend on the particular environment; we should probably do the same here for consistency.

eugeneius commented on guides/source/configuring.md:
```diff
@@ -401,6 +401,8 @@ The schema dumper adds one additional configuration option:
 
 * `config.action_controller.per_form_csrf_tokens` configures whether CSRF tokens are only valid for the method/action they were generated for.
 
+* `config.action_controller.default_protect_from_forgery` determines whether forgery protection is added on ActionController:Base.  This is false by default, but enabled when loading defaults for Rails 5.2.
```
ActionController:Base -> `ActionController::Base` (with backticks 😊)

eugeneius commented on actionpack/lib/action_controller/metal/request_forgery_protection.rb:
```diff
@@ -85,6 +85,10 @@ module RequestForgeryProtection
       config_accessor :per_form_csrf_tokens
       self.per_form_csrf_tokens = false
 
+      # Controls whether forgery protection is enabled by default
```
This line needs a full stop at the end.

eugeneius commented on actionpack/lib/action_controller/metal/request_forgery_protection.rb:
```diff
@@ -128,6 +132,15 @@ def protect_from_forgery(options = {})
         append_after_action :verify_same_origin_request
       end
 
+      # Turn off request forgery protection. This is a wrapper for
```
I think there should be a colon here:

> This is a wrapper for:

Commit:
Add ActionController::Base.skip_forgery_protection

Since we now default to `protect_from_forgery with: :exception`,
provide a wrapper to `skip_before_action :verify_authenticity_token`
for disabling forgery protection.

lugray commented:
Thanks @eugeneius, updated as per your review.

eugeneius commented:
This looks great! 👏 

My only remaining question is whether there should be an entry in [`new_framework_defaults_5_2.rb`](https://github.com/rails/rails/blob/3fa66935fd65e2d834dcc743bd835afb5b875f7f/railties/lib/rails/generators/rails/app/templates/config/initializers/new_framework_defaults_5_2.rb.tt) for this, so that upgraded apps can opt-in to the new behaviour if they want it. My understanding of that file is that it should mirror the options from `Rails::Application::Configuration#load_defaults`.

lugray commented:
@eugeneius That makes sense. I added to `new_framework_defaults_5_2.rb.tt`

UlrichThomasGabor commented on railties/lib/rails/generators/rails/app/templates/config/initializers/new_framework_defaults_5_2.rb.tt:
```diff
@@ -17,3 +17,7 @@
+
+# Add default protection from forgery to ActionController::Base instead of in
+# ApplicationController.
+# Rails.applocation.config.action_controller.default_protect_from_forgery = true
```
There is a typo `applocation`

lugray replied:
Thanks, I'll get a fix up.

lugray replied:
Actually, it already got fixed in e01b240e.

lugray commented:
@wnm By default, older versions do have it in `ApplicationController`.  The point I'm making is that the way to turn it off in older versions would have been to remove the line from the generated `ApplicationController`.  If someone has actively removed it because they don't want it, upgrading should not change the behavior.  But new applications will get the 5.2 defaults, which set it to true.  If you're upgrading an older version of rails, and you want to remove it from ApplocationController, you'll need to change the configuration you use, since you won't have the 5.2 defaults.

wnm commented:
@lugray ah I see... totally makes sense. thanks for the quick response!!

octavpo commented:
I think it would be better if the default_protect_from_forgery config could take the :with parameter as a value.

There's also an inconsistency in approach that if you use default_protect_from_forgery = true in config it defaults to :exception, while if you use protect_from_forgery with no parameters in the controller it defaults to :null_session.

collimarco commented:
> There's also an inconsistency in approach that if you use default_protect_from_forgery = true in config it defaults to :exception, while if you use protect_from_forgery with no parameters in the controller it defaults to :null_session.

I totally agree! It is not clear from the docs that the default behavior is `:exception`.

swrobel commented:
@lugray can you explain this?

> There's also an inconsistency in approach that if you use default_protect_from_forgery = true in config it defaults to :exception, while if you use protect_from_forgery with no parameters in the controller it defaults to :null_session.

rafaelfranca commented:
Yes, we change the default to be `:exception` because `:null_session` was a bad default.

swrobel commented:
@rafaelfranca that does not seem to be the case:
https://github.com/rails/rails/blob/master/actionpack/lib/action_controller/metal/request_forgery_protection.rb#L138
https://api.rubyonrails.org/classes/ActionController/RequestForgeryProtection/ClassMethods.html#method-i-protect_from_forgery

rafaelfranca commented:
It was the case already before this patch.  https://github.com/rails/rails/pull/29742/files#diff-8380e0e11d4efeb2b0b142ae9e4cf512L3 Even that the default of the method is null_session, rails already generated the application using exception. That was the new default for applications not for the method.

rafaelfranca commented:
I think it is worth considering it, but it will be a really annoying experience to most of the users given they will have to be explicit about the default to remove the deprecation warning. But I guess it will avoid confusion.
